### PR TITLE
trace2:gvfs:experiment Add experimental regions and data events to help diagnose checkout and reset perf problems

### PR DESCRIPTION
**Pull request by jeffhostetler**
Add experimental regions and data events to help diagnose
performance problems in checkout and reset.

We may or may not choose to carry these changes forward in the future.

### CODE REVIEW

File: packfile.c [hunk 1]
```diff
@@ -1646,6 +1646,12 @@ static void *read_object(struct repository *r,
 	return content;
 }
 
+static unsigned long g_nr_unpack_entry;
+unsigned long get_nr_unpack_entry(void)
```
Comment (derrickstolee) [inline packfile.c]: This needs to be static if it's not in the header file.

Comment (derrickstolee) [reply packfile.c]: Or marked `extern` in this file.

Comment (jeffhostetler) [reply packfile.c]: got it.  thanks!  i was trying to not pollute the header file with an experimental function and then have to include it into all of the callers.

**Review by derrickstolee — APPROVED** [review]

**Comment by derrickstolee** [comment]
@jeffhostetler looks like t9800 needs an adjustment to its check for tracing, which is annoying.

File: unpack-trees.c [hunk 2]
```diff
@@ -1660,6 +1663,8 @@ int unpack_trees(unsigned len, struct tree_desc *t, struct unpack_trees_options
 done:
 	trace_performance_leave("unpack_trees");
 	clear_exclude_list(&el);
+	trace2_data_intmax("exp", NULL, "unpack_trees/nr_unpack_entries",
```
Comment (dscho) [inline unpack-trees.c]: It seems that sometimes, the experiments are prefixed with `"exp"`, but sometimes not. Did you mean to prefix all of them in that manner, or was it intentional to do that only with some, but not others?

**Review by dscho — APPROVED** [review]

**Comment by dscho** [comment]
> looks like t9800 needs an adjustment to its check for tracing, which is annoying.

I don't know... that's a p4 test that seems to be flaky. I restarted the `linux-gcc` job; the `windows-build` job is broken because I had forgotten to open #162.

**Comment by derrickstolee** [comment]
```

    CC credential.o
config.c: In function 'git_parse_signed':
config.c:873:10: error: absolute value function 'labs' given an argument of type 'intmax_t' {aka 'long long int'} but has parameter of type 'long int' which may cause truncation of value [-Werror=absolute-value]
  873 |   uval = labs(val);
      |          ^~~~
config.c:875:21: error: absolute value function 'labs' given an argument of type 'intmax_t' {aka 'long long int'} but has parameter of type 'long int' which may cause truncation of value [-Werror=absolute-value]
    CC csum-file.o
    CC ctype.o
  875 |   if (uval > max || labs(val) > uval) {
      |                     ^~~~
    CC date.o
cc1.exe: all warnings being treated as errors
    CC delta-islands.o
make: *** [Makefile:2383: config.o] Error 1
make: *** Waiting for unfinished jobs....
    CC decorate.o
```
so many cross-plat things to keep in mind. :(

File: unpack-trees.c (hunk 2 above)
Comment (jeffhostetler) [reply unpack-trees.c]: yeah, i've been inconsistent.  i was using "exp" for things i consider really temporary while we gather some data for a release or 2.  most of the other ones that have real categories, but have "experiment" in the commit message are things that i could keep and maybe upstream or could just drop.  i'm trying to not flood the telemetry stream with things that may not be useful long term, but at the same time gather enough data for our immediate analysis needs.

**Comment by dscho** [comment]
> ```
>
>     CC credential.o
> config.c: In function 'git_parse_signed':
> config.c:873:10: error: absolute value function 'labs' given an argument of type 'intmax_t' {aka 'long long int'} but has parameter of type 'long int' which may cause truncation of value [-Werror=absolute-value]
>   873 |   uval = labs(val);
>       |          ^~~~
> config.c:875:21: error: absolute value function 'labs' given an argument of type 'intmax_t' {aka 'long long int'} but has parameter of type 'long int' which may cause truncation of value [-Werror=absolute-value]
>     CC csum-file.o
>     CC ctype.o
>   875 |   if (uval > max || labs(val) > uval) {
>       |                     ^~~~
>     CC date.o
> cc1.exe: all warnings being treated as errors
>     CC delta-islands.o
> make: *** [Makefile:2383: config.o] Error 1
> make: *** Waiting for unfinished jobs....
>     CC decorate.o
> ```
> so many cross-plat things to keep in mind. :(

Indeed. Sorry about my mistake: I should have fixed this in
microsoft/git a lot earlier...

**Pull request merged**